### PR TITLE
feat(gatsby-plugin-feed): Allows to override default feed url generation

### PR DESCRIPTION
**Pull request by matyo91**
## Description
This add the possibility to configure a custom link for rss meta generation.

### CODE REVIEW

File: packages/gatsby-plugin-feed/src/gatsby-ssr.js [hunk 1]
```diff
@@ -16,9 +16,15 @@ exports.onRenderBody = ({ setHeadComponents, pathname }, pluginOptions) => {
...
+    .map(({ output, title, link }, i) => {
+      let href = output
+      if(link) {
+        href = link
+      } else {
+        if (output.charAt(0) !== `/`) {
+          href = `/` + output
+        }
+        href = withPrefix(href)
       }
```
Comment (muescha) [inline packages/gatsby-plugin-feed/src/gatsby-ssr.js]: just as suggestion: maybe with *"little magic"* it can be shorter:

```javascript
let href = link || withPrefix(output.replace(/^\/?/,`/`))
```

Comment (pvdz) [reply packages/gatsby-plugin-feed/src/gatsby-ssr.js]: Hm, that magic does something with a regular expression that did not need a regular expression before.

I can only see the diff part in this comment but looking at that I would prefer the old style tbh. It's better for perf and legibility.

One alternative might be to use `!output.startsWith('/')`, but that'd be a nit.

File: packages/gatsby-plugin-feed/src/gatsby-ssr.js [hunk 2]
```diff
@@ -16,18 +16,16 @@ exports.onRenderBody = ({ setHeadComponents, pathname }, pluginOptions) => {
       if (typeof match === `string`) return new RegExp(match).exec(pathname)
       return true
     })
-    .map(({ output, title }, i) => {
-      if (output.charAt(0) !== `/`) {
-        output = `/` + output
-      }
+    .map(({ output, title, link }, i) => {
+      const href = link || withPrefix(output.replace(/^\/?/, `/`))
```
Comment (pvdz) [inline packages/gatsby-plugin-feed/src/gatsby-ssr.js]: Aside from the other comment, when I look at `withPrefix` it is actually `withAssetPrefix` or `withPrefix` from `gatsby-link/src/index.js`. Looking at that I see the following for these functions:

```js
export function withPrefix(path) {
  return normalizePath(
    [
      typeof __BASE_PATH__ !== `undefined` ? __BASE_PATH__ : __PATH_PREFIX__,
      path,
    ].join(`/`)
  )
}

export function withAssetPrefix(path) {
  return [__PATH_PREFIX__].concat([path.replace(/^\//, ``)]).join(`/`)
}

function normalizePath(path) {
  return path.replace(/\/+/g, `/`)
}
```

In both of these cases it'll apply the fix applied here as well. So shouldn't this just be `withPrefix(output)`? Or am I missing something?

( 
_Hooking back to my other comment_; 
This is why I don't like the regexes in this context; there's three here that do overlapping-but-not-the-same things. Regular code would have made their difference more obvious.

```js
s.replace(/^\/?/, `/`)
s.replace(/^\//, ``)
s.replace(/\/+/g, `/`)
```

So 
- the new regex does a "if the string does not start with `/`, add it, otherwise noop"
- the second regex strips the leading `/` (making the new regex pointless)
- the third regex replaces all repeating forward slashes with a single slash

While the third is different from the new regex, the code will do a `join('/')` before reaching that point which either leads to a single or double forward slash, so effectively making the new regex pointless as well.
)

In other words. Please call `withPrefix(output)`, without forward slash scrubbing :)

File: packages/gatsby-plugin-feed/src/gatsby-ssr.js (hunk 1 above)
Comment (pvdz) [reply packages/gatsby-plugin-feed/src/gatsby-ssr.js]: (Nevermind, you can just drop that part altogether)

File: packages/gatsby-plugin-feed/src/gatsby-ssr.js (hunk 2 above)
Comment (matyo91) [reply packages/gatsby-plugin-feed/src/gatsby-ssr.js]: The only remark i see here is `// TODO: remove for v3` : https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-plugin-feed/src/gatsby-ssr.js#L5

If i continue your analyse, it seem's that previously the author allowed miss-typing into `output` and I presume the same for `__BASE_PATH__` and `__PATH_PREFIX__ `.

so what you suggest is `withPrefix(output)`, but it may be not the final solution, as you are not allowing miss-typing to `output` with was the case before.

so https://github.com/gatsbyjs/gatsby/pull/19869#discussion_r351969581 from @muescha was also a good suggestion in case of non-regression.

Comment (muescha) [reply packages/gatsby-plugin-feed/src/gatsby-ssr.js]: I also looked into this 2 possible functions and but then decided not change too much existing code 

There only 2 combinations possible: 
- RegEx1 + RegEx2 + Join(/) or 
- RegEx1 + join(/) + RegEx3

But if now 3x2 eyes looking on it I would also follow the less code suggestion from @pvdz

Comment (matyo91) [reply packages/gatsby-plugin-feed/src/gatsby-ssr.js]: I prefer cut the conversation in favor of non-regression context. Thanks !

**Comment by matyo91** [comment]
> With the test you can drop the (redundant) regular expression safely without having to worry about v3 :)

I added more tests one for the feature and two for prefixes.
I made tests about what is done today for non-regression, not what V3 will be tomorrow.

So i didn't understand why I can in this PR drop regular expression and potentially break other projets using this plugin, as I have no idea if this PR will be released for V2 or for V3.
